File: Week8 Project 3-5/src/com/rthr/cs/WageCalculator.java
/**
 * Author:  Arthur Lockman
 * Date:    Oct 24, 2012
 * School:  Gorham High School
 * 
 * Project: Week8 Project 3-5
 * Package: com.rthr.cs
 */
package com.rthr.cs;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

/**
 * Calculates the pay of an employee.
 * 
 * @author arthurlockman
 */
public class WageCalculator
{
	private static FileInputStream inFile;
	private static InputStreamReader inReader;
	private static BufferedReader reader;
	
	private static String employeeName;
	private static double hourlyWage = 0;
	private static double regularHours = 0;
	private static double overtimeHours = 0;
	
	/**
	 * Calculates the pay of an employee. Prompts the user
	 * for all needed data to do the calculations.
	 * @param args Program arguments (none).
	 * @throws IOException 
	 */
	public static void main(String[] args) throws IOException
	{
		WageCalculator wCalc = new WageCalculator();
		wCalc.run();
	}
	
	/**
	 * Runs the wage calculations.
	 * @throws IOException 
	 */
	public void run() throws IOException
	{
		
		//Collect data.
		this.initFile("/Users/arthurlockman/Dropbox/School/AP CS/Work/Data Files/VHSP35data1_1.txt/");
		this.parseData();
		
		//Return data.
		double totalPay = this.findTotalPay(hourlyWage, regularHours, overtimeHours);
		System.out.println("Employee "+employeeName+" earned $"+totalPay+" this week.");
		
		System.exit(0);
	}
	
	/**
	 * Prepares the file for input.
	 * @param filePath the path to the file.
	 */
	private void initFile(String filePath)
	{
		try
		{
			inFile = new FileInputStream(filePath);
			inReader = new InputStreamReader(inFile);
			reader = new BufferedReader(inReader);
		} catch (FileNotFoundException e)
		{
			e.printStackTrace();
			System.exit(1);
		}
	}
	
	/**
	 * Parses the data in the data file.
	 * @throws IOException 
	 */
	private void parseData() throws IOException
	{
		String data = reader.readLine();
		StringTokenizer tokenizer = new StringTokenizer(data);
		
		employeeName = tokenizer.nextToken()+" "+tokenizer.nextToken();
		hourlyWage = Double.parseDouble(tokenizer.nextToken());
		
		for (int i=0; i<tokenizer.countTokens(); i++)
		{
			regularHours = regularHours + Double.parseDouble(tokenizer.nextToken());
			overtimeHours = overtimeHours + Double.parseDouble(tokenizer.nextToken());
		}
	}
	
	/**
	 * Calculates the total pay of the employee.
	 * 
	 * @param hourlyWage The employee's hourly wage.
	 * @param regularHours The total number of regular hours worked by the employee.
	 * @param overtimeHours The total number of overtime hours worked by the employee.
	 * @return A double, the total pay of the employee.
	 */
	private double findTotalPay(double hourlyWage, double regularHours, double overtimeHours)
	{
		return this.round2(hourlyWage * regularHours + hourlyWage * 1.5 * overtimeHours);
	}

	/**
	 * Rounds a double to two decimal places.
	 * @param num The number to round.
	 * @return A double, rounded to two decimal places.
	 */
	private double round2(double num)
	{
		double num1 = num*100;
		num1 = num1+.5;
		int num2 = (int)num1;
		num1 = (double)num2;
		num1 = num1/100;
		return num1;
	}
}

File: Week8 Project 3-5 Xcode/Week8 Project 3-5 Xcode/WageCalculator.m
//
//  WageCalculator.m
//  Week8 Project 3-5 Xcode
//
//  Created by Arthur Lockman on 10/24/12.
//  Copyright (c) 2012 Arthur Lockman. All rights reserved.
//

#import "WageCalculator.h"

@implementation WageCalculator

@synthesize hourlyWage, overtimeHours, regularHours, data, employeeName;

/**
 Run the program.
 */
- (void) run
{
	[self openFile:@"/Users/arthurlockman/Dropbox/School/AP CS/Work/Data Files/VHSP35data1_1.txt"];
	[self parseData];
	NSLog(@"Employee %@ earned %@ this week.",employeeName,[self round2:[self calculatePay] andFormat:NSNumberFormatterCurrencyStyle]);
}

/**
 Opens and reads in the data from the data file.
 @param filePath The path to the file on the filesystem.
 */
- (void) openFile:(NSString *)filePath
{
	NSError *fOpenError;
	data = [[NSString alloc] initWithContentsOfFile:filePath
										   encoding:NSUTF8StringEncoding error:&fOpenError];
	if (data == nil)
	{
		NSLog(@"%@",[fOpenError localizedFailureReason]);
		exit(1);
	}
}

/**
 Parses the data from the data file.
 */
- (void) parseData
{
	NSArray* parsedData = [data componentsSeparatedByString:@" "];
	employeeName = [NSString stringWithFormat:@"%@ %@",parsedData[0],parsedData[1]];
	hourlyWage = [parsedData[2] doubleValue];
	
	for (int i=3; i<12; i=i+2)
	{
		regularHours = regularHours + [parsedData[i] doubleValue];
		overtimeHours = overtimeHours + [parsedData[i+1] doubleValue];
	}
}

/**
 Calculates the employee's pay.
 @return The pay for the employee this week.
 */
- (double) calculatePay
{
	return hourlyWage * regularHours + hourlyWage * 1.5 * overtimeHours;
}

/**
 Round a double to two decimal places.
 @param number The number to round.
 @param style The style to format the number in.
 @return An NSString, the formatted rounded number.
 */
- (NSString*) round2:(double)number andFormat:(NSNumberFormatterStyle)style
{
	NSNumberFormatter* formatter = [NSNumberFormatter new];
	[formatter setNumberStyle:style];
	NSString* num = [formatter stringFromNumber:[NSNumber numberWithDouble:number]];
	return num;
}

@end
